### ISSUE TITLE
Custom setting cannot be evaluated in test_package

### ISSUE THREAD

**Issue by olivren** [conan-io/conan#1427]
I want to introduce a new setting, along with `os`, `arch`, `compiler` and `build_type`, called `product`. So, I added this line at the end of `~/settings.yml`.

    product: [gabu, zomeu]

I then added the new setting as a string in the `settings` attribute, and try to get its value using `self.settings.product`. The problem is that this works fine except in the conan file of the test_package.

conanfile.py:
```
from conans import ConanFile

class MyLibConan(ConanFile):
    name = 'mylib'
    version = '1.0'
    settings = 'os', 'product'
    
    def requirements(self):
        print 'In main recipe:'
        print 'os: ', self.settings.os
        print 'product: ', self.settings.product
```

test_package/conanfile.py:
```
from conans import ConanFile

class MyLibConanTest(ConanFile):
    requires = 'mylib/1.0@me/stable'
    settings = 'os', 'product'
    
    def requirements(self):
        try:
            print 'In test_package recipe:'
            print 'os: ', self.settings.os
            print 'product: ', self.settings.product
        except Exception as e:
            print 'exception!! ', e
        
    def test(self):
        print 'ok'
```

When invoked as `conan test_package -s product=gabu`, I get this output (error is on line 4):
```
C:\tmp\conan_bug_custom_setting>conan test_package -s product=gabu
In test_package recipe:
os:  Windows
product:  exception!!  __str__ returned non-string (type NoneType)
Exporting package recipe
mylib/1.0@me/stable: The stored package has not changed
In test_package recipe:
os:  Windows
product:  gabu
In main recipe:
os:  Windows
product:  gabu
Requirements
    mylib/1.0@me/stable from local
Packages
    mylib/1.0@me/stable:8379b017433f1db75d73bef3f369980554634341

mylib/1.0@me/stable: WARN: Forced build from source
mylib/1.0@me/stable: Building your package in C:\Users\me\.conan\data\mylib\1.0\me\stable\build\8379b017433f1db75d73bef3f369980554634341
mylib/1.0@me/stable: Copying sources to build folder
mylib/1.0@me/stable: Generated txt created conanbuildinfo.txt

mylib/1.0@me/stable: Package '8379b017433f1db75d73bef3f369980554634341' built
mylib/1.0@me/stable: Build folder C:\Users\me\.conan\data\mylib\1.0\me\stable\build\8379b017433f1db75d73bef3f369980554634341
mylib/1.0@me/stable: Generated conaninfo.txt
mylib/1.0@me/stable: Generated conanbuildinfo.txt
mylib/1.0@me/stable: Generated conanenv.txt
mylib/1.0@me/stable: Generating the package
mylib/1.0@me/stable: Package folder C:\Users\me\.conan\data\mylib\1.0\me\stable\package\8379b017433f1db75d73bef3f369980554634341
mylib/1.0@me/stable: WARN: This conanfile has no package step
mylib/1.0@me/stable package(): WARN: No files copied!
mylib/1.0@me/stable: Package '8379b017433f1db75d73bef3f369980554634341' created
PROJECT: Generated txt created conanbuildinfo.txt
PROJECT: Generated txt created conanbuildinfo.txt
PROJECT: Generated conaninfo.txt
Project: WARN: This conanfile has no build step
ok
```

Note that here I catch the exception, but if I do not then I end up with an obscure message `ERROR: Unable to retrieve first requirement of test conanfile.py`. Also note that the value of `self.settings.os` is correctly evaluated.

**Comment by lasote** [comment]
I cannot reproduce the error, I'm getting the `product` setting printed ok. Why are you using it in the requirements of the test_package? Maybe the configure is a better place or you are trying to do something with additional requires?

**Comment by lasote** [comment]
I don't think it is the error but try to quote the settings values in the setting.yml to discard:

product: ["gabu", "zomeu"]

**Comment by memsharded** [comment]
I have been able to reproduce it, indeed something seems very weird. Investigating.

**Comment by memsharded** [comment]
There is a bug, ``settings`` are not being set before the ``requirements()`` method is called in test_package().

But, once you have captured that Exception, the rest should work fine, so I don't know what could be happening. Maybe it could be a good idea to put exactly the above in a github repo, so we can make sure that we are trying exactly the same code, would you mind creating it, please?

Also, if you can please asses about the urgency of this bug: do you really need a ``requirements()`` method in your test_package/conanfile? Do you have conditional requirements in your test_packages? Please tell. Thanks very much for your feedback.

**Comment by memsharded** [comment]
I am working on a fix. In the meantime, you might workaround this bug by adding the desired product in your ``conan.conf`` defaults:

```
[settings_defaults]
product:  gabu

**Comment by memsharded** [comment]
Ok, so not necessary to setup a repo with the code, I have already integrated a test in the code base (and the fix, of course): https://github.com/conan-io/conan/pull/1429

It will be in next 0.25 release. Thanks for the help!

**Comment by olivren** [comment]
Thanks for the very quick feedback ! I was indeed trying to have different 
requirements in the test_package recipe depending on this new setting. But 
it turns out it was not necessary for what I wanted to achieve, so this bug 
is not a blocker for me. I submitted the bug anyway, because it's clearly 
not the intended behavior.

**Comment by memsharded** [comment]
Excellent thank you!
Happy to know that it is not a blocker.

**Comment by memsharded** [comment]
It has been fixed in the develop branch, will be in next release.

**Comment by memsharded** [comment]
Released in 0.25
